src/FuseGPUThreadLoops.cpp

```diff
@@ -359,19 +363,19 @@ class ExtractSharedAndHeapAllocations : public IRMutator {
                         get_compiler_logger()->record_non_monotonic_loop_var(op->name, s.size);
                     }
                     precompute_allocation_size(s);
-                } else {
-                    auto interval_bounds = bounds_of_expr_in_scope(s.size, scope);
-                    user_assert(interval_bounds.has_upper_bound())
-                        << "Couldn't infer bounds for " << s.name << " shared memory allocation\n";
-                    // In theory we could precompute the allocation
-                    // size if there's no upper bound too, but for the
-                    // assert above to fail we'd have to encounter an
-                    // expression that is_monotonic detects as
-                    // increasing, decreasing, or constant, but is
-                    // somehow unbounded. It's probable that no such
-                    // expression exists. is_monotonic is generally
-                    // less capable than bounds_of_expr_in_scope.
-                    s.size = interval_bounds.max;
+                    break;
+                case Monotonic::Increasing:
+                    s.size = substitute(op->name, simplify(op->min + op->extent - 1), s.size);
+                    break;
+                case Monotonic::Constant:
+                    // The size expression used the variable, but we
```

Comment:
In which cases is the dependence false somehow?

Reply:
Or I guess, if there are cases, is that a bug we should be worried about?

Reply:
@abadams added this code, I am not sure, Andrew?

Reply:
expr_uses_var occurs before solve/simplify/cse, so the most common reason it might hit this path is just that the var was successfully removed by the simplifier. But I thought it was also possible that is_monotonic might return Constant, even if the var appears in the expr. I didn't want to rule out the possibility that is_monotonic might become more powerful than the simplifier's ability to cancel things in some obscure cases. An example might be something dumb like `x - likely(x)`, where you can replace `likely` with anything that's an identify function. Or perhaps an appearance of x in a place that doesn't affect the derivative, e.g: `print(37, x)`. Or maybe one day is_monotonic might realize that `select(x < 0, 0, 1) + select(0 <= x, 1, 0)` doesn't actually depend on x but the simplifier won't.

Reply:
Ah gotcha, that makes sense.